modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/fsm/QueryExecutor.java

```diff
@@ -241,6 +254,229 @@ CompletableFuture<AsyncSqlCursor<InternalSqlRow>> executeChildQuery(
+                                        firstCursor = currentCursor;
+                                    }
+
+                                    currentCursor.onClose().thenRun(() -> query.moveTo(ExecutionPhase.TERMINATED));
+                                }
+
+                                return completedFuture(firstCursor);
+                            })
+                            .thenCompose(Function.identity());
+                });
+    }
+
+    private CompletableFuture<AsyncSqlCursor<InternalSqlRow>> executeSequentially(List<Query> queries) {
+        CompletableFuture<AsyncSqlCursor<InternalSqlRow>> firstCursor = null;
+        CompletableFuture<AsyncSqlCursor<InternalSqlRow>> cursorRef = null;
+        CompletableFuture<Void> lastStep = nullCompletedFuture();
+        for (Query query : queries) {
+            query.reset();
+
+            CompletableFuture<AsyncSqlCursor<InternalSqlRow>> cursorRef0 = cursorRef;
+            CompletableFuture<AsyncSqlCursor<InternalSqlRow>> cursorFuture = lastStep
+                    .whenComplete((none, ex) -> {
+                        if (ex != null) {
+                            query.onError(ex);
+                        }
+                    })
+                    .thenCompose(none -> Programs.SCRIPT_ITEM_EXECUTION.run(query)
+                            .whenComplete((cursor, ex) -> {
+                                if (cursorRef0 != null) {
+                                    if (cursor != null) {
+                                        cursorRef0.complete(cursor);
+                                    } else {
+                                        cursorRef0.completeExceptionally(ex);
+                                    }
+                                }
+
+                                if (cursor != null) {
+                                    cursor.onClose().thenRun(() -> query.moveTo(ExecutionPhase.TERMINATED));
+                                }
+                            }));
+
+            lastStep = cursorFuture.thenCompose(AsyncDataCursor::onFirstPageReady);
+            cursorRef = query.nextCursorFuture;
+
+            if (firstCursor == null) {
+                firstCursor = cursorFuture;
+            }
+        }
+
+        return firstCursor;
+    }
+
+    AsyncSqlCursor<InternalSqlRow> createAndSaveSqlCursor(Query query, AsyncDataCursor<InternalSqlRow> dataCursor) {
+        QueryPlan plan = query.plan;
+
+        assert plan != null;
+
+        AsyncSqlCursorImpl<InternalSqlRow> cursor = new AsyncSqlCursorImpl<>(
+                plan.type(),
+                plan.metadata(),
+                dataCursor,
+                query.nextCursorFuture
+        );
+
+        query.cursor = cursor;
+
+        query.cancel.add(timeout -> dataCursor.cancelAsync(
+                timeout ? CancellationReason.TIMEOUT : CancellationReason.CANCEL
+        ));
+
+        return cursor;
+    }
+
+    static class ParsedResultWithNextCursorFuture {
```

Comment:
Maybe it's worth moving the nested class declaration to the end of the class?

Reply:
make sense, moved o the very bottom of the class